File: stack/typescript/Stack.ts
interface IStack<T> {
  arr: T[];
  top: number;
  capacity: number;

  getSize: () => number;
  isEmpty: () => boolean;
  isFull: () => boolean;
  push: (value: T) => void;
  pop: () => void;
  peek: () => T;
  getStack: () => T[];
  showStack: () => void;
}


export class Stack<T> implements IStack<T> {
  arr: T[];
  top: number;
  capacity: number;
  
  constructor(capacity: number) {
    this.arr = [];
    this.top = -1;
    this.capacity = capacity;
  }

  getSize(): number {
    return this.top > 0 ? this.top + 1 : 0
  }

  isEmpty(): boolean {
    return this.top < 0;
  }

  isFull(): boolean {
    return this.top >= this.capacity;
  }

  push(value: T): void {
    if (this.isFull()) {
      console.log("El Stack esta lleno.")
      return;
    }
    console.log("Insertando ", value);
    this.arr[++this.top] = value;
  }

  pop(): void {
    if (this.isEmpty()) {
      console.log("El Stack está vacío.")
      return;
    }
    else {
      console.log("Eliminando el item en la posición: ", this.top);
      this.arr.splice(this.top, 1);
      this.top--;
    }
  }

  peek(): T {
    return this.arr[this.top];
  }

  getStack(): T[] {
    let output: T[] = [];
    for (let i = 0; i < this.arr.length; i++) {
      output.push(this.peek());
      this.top--;
    }
    return output;
  }

  showStack(): void {
    console.log("\nEstado Actual del Stack: \n")
    for (let i = this.capacity; i > this.arr.length; i--) {
      console.log("| " + " " + " |")
    }
    for (let i = this.arr.length - 1; i >= 0; i--) {
      if (i === this.top) {        
        console.log("| " + this.arr[i] + " |" + " -> " + " TOP")
      }
      else {
      console.log("| " + this.arr[i] + " |")
      }
    }
    console.log("_____\n")
  }
}

File: stack/python/Stack.py
class Stack:
  def __init__(self, capacity) -> None:
    self.arr = []
    self.top = -1
    self.capacity = capacity
  
  def get_size(self) -> bool:
    return self.top + 1 if self.top > 0 else 0
  
  def is_empty(self) -> bool:
    return self.top < 0
  
  def is_full(self) -> bool:
    return self.top >= self.capacity - 1

  def push(self, value) -> None:
    if (self.is_full()):
      print("El Stack está lleno.")
      return None
    self.arr.append(value)
    self.top += 1
    self.showStack()
  
  def pop(self) -> None:
    if self.is_empty():
      print("El Stack está vacío.")
      return None
    del self.arr[self.top]
    self.top -= 1
    self.showStack()
    
  def showStack(self) -> None:
    print("\nDiagrama del Stack: \n")
    for i in range(len(self.arr), self.capacity):
      print("\t|  " + " " + "  |")
    
    j = len(self.arr) - 1
    while(j >= 0):
      if j == self.top:
        print("\t| ", self.arr[j],  " |" + " -> " + " TOP")
      else:
        print("\t| ", self.arr[j],  " |")
      j -= 1
    print("\t|_____|")